### DISCUSSION TITLE
Avoid weak master key

### DISCUSSION THREAD

**Discussion by gmourier** [meilisearch/product#538] [Feedback & Feature Proposal]
Sub-discussion for https://github.com/meilisearch/product/discussions/536

## Description

Some time ago, @brunoocasali initiated a [product discussion](https://github.com/meilisearch/product/discussions/443) regarding the master key behavior to naturally guide users to a secured usage of Meilisearch.

Since the @meilisearch/cloud-team uses the master key for all operations, we have been unable to make the changes that initially seemed to help strengthen security.

However, we could consider introducing a more demanding format to avoid users setting a weak master key like `12345678` or `masterKey` to add more security guidance.

**Comment by curquiza** [comment]
I personally like this idea
- it's easy to implement
- it's a "free" security addition: forcing for 8 characters for example is not asking too much from the users, and is for their safety
- it's really breaking so we will not be able to do it after v1

**Comment by brunoocasali** [comment]
> However, we could consider introducing a more demanding format to avoid users setting a weak master key like 12345678 or masterKey to add more security guidance.

So the idea is to follow a particular format or just to ensure a minimum hash size?

**Reply by gmourier** [reply]
Good question @brunoocasali.

A minimum hash size would be less secure `e.g 00000000` than a particular format but at least more secure than today.  
On the other side, a particular format needs more explanations and can be difficult to generate.

We could provide tips to generate a master-key in the error message if the master key is not valid, plus explain how to generate it the same way in the docs. WDYT?

**Reply by brunoocasali** [reply]
I like the idea of having a particular format + minimum hash size. 

We just want to ensure the users know clearly what they need to do to provide a good masterKey.

CC: @meilisearch/integration-team I know this would represent a very good amount of work for us, but in any case this brings more security to our software, not easy to decide but I would like your input here as well ;)

**Reply by curquiza** [reply]
I'm not sure this is a lot of work with our automation 🤔

**Reply by bidoubiwa** [reply]
I think asking for a longer master key is already a good security measure, a good balance between security and ease of setup. Are there other databases that are requiring password to be formatted like that?

**Reply by brunoocasali** [reply]
Agreed, but besides this point, we also have to think about how the user will face this problem and how they will solve the problem when it comes.

For example, will they have to create this token by themselves? Or will we provide a CLI option to generate one for them? If they try their own-generated token, which error and how will they understand the problem, Meilisearch will start anyway? 

I think this is something it will seriously impact the first contact users have with us...

**Reply by gmourier** [reply]
> Are there other databases that are requiring password to be formatted like that?

Not to my knowledge but it's something that is done by database operators for sure when exposing it. e.g https://kb.iu.edu/d/aphg

Although this can be thought out and done by experienced users by hand, the idea here is to **ensure** that a beginner can expose a Meilisearch endpoint on the public web under optimal security conditions.

**Enforcing a more secure format for the master key will only solve some flaws.** 

We have seen users exposing the master key on the frontend. Solving that question from the engine side has been delayed due to the fact that the [cloud-team currently uses the master-key to make every call on a Meilisearch instance](https://github.com/meilisearch/product/discussions/443#discussioncomment-2862683). When they will look at the API key management UX, we must take that opportunity to consider fixing this flaw. Signal 📡 to @davelarkan.

If the master key is not exposed, having a more secure format protects the master key from being brute-forced by a malicious attempt.

**Comment by gmourier** [comment]
I would like to see this considered for v1.0

It means preventing users from using what we could consider a weak master key and guiding them most efficiently not to tarnish their experience, as @brunoocasali mentioned.

If you have solution proposals, we can start discussing them here.

For now, @brunoocasali proposed to expect a particular format + a minimum hash size.

**Reply by bidoubiwa** [reply]
I'm more in favor of only a minimum hash-size to avoid additional complexity in starting a Meilisearch. If the user is brute-forcing the password, having a particular hash will only increase the time to find it. Is it worth the additional complexity?

**Reply by brunoocasali** [reply]
If we have a smooth way to propose a secure hash to the user, I think it does not matter if the token should be alphanumeric or not (or any other policy).

We can propose a workflow like this:

- User tries to start meilisearch with a masterKey = `masterKey` (unsafe token because it does not have numbers nor enough characters, ex. min is 24). 
- Meilisearch proposes a new one to the user:

```bash
➜  meilisearch-ruby git:(main) ./meilisearch

888b     d888          d8b 888 d8b                                            888
8888b   d8888          Y8P 888 Y8P                                            888
88888b.d88888              888                                                888
888Y88888P888  .d88b.  888 888 888 .d8888b   .d88b.   8888b.  888d888 .d8888b 88888b.
888 Y888P 888 d8P  Y8b 888 888 888 88K      d8P  Y8b     "88b 888P"  d88P"    888 "88b
888  Y8P  888 88888888 888 888 888 "Y8888b. 88888888 .d888888 888    888      888  888
888   "   888 Y8b.     888 888 888      X88 Y8b.     888  888 888    Y88b.    888  888
888       888  "Y8888  888 888 888  88888P'  "Y8888  "Y888888 888     "Y8888P 888  888

Database path:        "./data.ms"
Server listening on:  "http://0.0.0.0:7700"
Environment:          "development"
Commit SHA:           "unknown"
Commit date:          "unknown"
Package version:      "0.29.1"
Anonymous telemetry:  "Disabled"

You used a WEAK master key. To run meilisearch with a master key, you should provide a secure master key containing (letters and numbers with 24 characters minimum).
We generated a secure key for you (you can safely copy this token): 

>>> T0d6Ddx6A1y36CdP75c1fDF5 <<<


Documentation:        https://docs.meilisearch.com
Source code:          https://github.com/meilisearch/meilisearch
Contact:              https://docs.meilisearch.com/resources/contact.html

[2022-11-17T20:26:49Z ERROR  ........
```

- Then, the user gets the generated token and uses it to start Meilisearch. If they try again with the same weak token, we should send a different key to the user.

It is not good to prevent users from initializing the instance with an error like this, but at least the error and the message could be helpful enough for the user quickly fix it without even going to the docs.

Also, we can create a CLI option that generates the token for the user, something like:

`./meilisearch --generate-master-key`

**Reply by bidoubiwa** [reply]
I like your solution, in both cases, with or without formating. I'm still more in favor of no-formating but just a minimum length string. Nonetheless, if we do go for your suggestion, your solution makes it user-friendly enough not to frustrate them too much.

**Reply by gmourier** [reply]
I also like @brunoocasali's solution. I would like the input from someone in the engine team before we go down this road.

@dureuill, I would love your input and suggestion on that whenever you have time 🙇‍♂️ thank you!

**Reply by dureuill** [reply]
Hey @gmourier, apologies for the delay in responding to this, I'm still unsure of the broader security context of API keys (such as the attacker model and guaranteed properties).

I think I generally like @brunoocasali's solution, I can think of several pieces of software with a similar behavior:

- Jupyter Notebook [generates a token](https://jupyter-notebook.readthedocs.io/en/stable/security.html) to use in URLs for authentication on each startup. Because it is done by default, it can be a bit confusing and aggravating to unsuspecting users in a context where security is not felt as necessary.  So even though it wasn't a proposal I like the fact that we can currently use meilisearch in "dev mode" without a master key at all.
- The Rocket web framework [requires a secret key in release mode but generates it in debug mode](https://rocket.rs/v0.5-rc/guide/configuration/#secret-key). It also specifies the accepted format which I think would be reasonable for us too. 

If we do go with @brunoocasali's route, it would be great to advertise the `--generate-master-key` in case the software is run without a master key.

**Reply by brunoocasali** [reply]
This is related to the central issue: https://github.com/meilisearch/meilisearch-dotnet/issues/363#issuecomment-1332593433

**Reply by dureuill** [reply]
Thank you for the additional context, from there I can gather we'd need at least 16 bytes for the key... Do we know if we have other known constraints on the key's format? I think it is OK to impose constraints on the keys if we are providing a way to generate them.

Thinking more about key generation, I wonder if we should provide the feature itself (in the form of a flag like discussed), or if we could limit ourselves to outputting commands for users to execute on their machine, freeing us of the burden of providing cryptographically secure key generation. What do you think?

**Reply by brunoocasali** [reply]
What do you mean by `outputting commands for users to execute on their machine`? Like accessing some system library to generate the key?

**Reply by dureuill** [reply]
No I was thinking giving the command lines for the user to execute, much like GitHub is doing with SSH keys: https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent#generating-a-new-ssh-key

But let's disregard this, we should be able to generate secure master keys using a cryptographically secure random number generator.

**Comment by gmourier** [comment]
Thanks, everyone; the proposed ideas seem to solve the weak master key issue. To clarify, here is where we are exploring this topic, it also raises some questions we want to consider.

## Enforcing a minimum number of chars for the master key instead of a specific format (e.g. uuid-v4, sha256, etc)

A suggestion brought up by @dureuill during the engine weekly meeting; 

We don't necessarily want to wait for a specific format. Asking for a minimum number of characters could be sufficient to guide users to a more secure master key without requiring a format that can be difficult to generate. (@dureuill, please correct me if it's unclear or does not relate to your suggestion). 

WDYT? 

---

### How many minimum characters should Meilisearch expect for the master key to be valid?

- `8` was mentioned by @curquiza
- `16` was mentioned by @dureuill

IMO, if no particular format is requested other than the current one (The master key must be composed of valid utf-8 characters), 16 characters seem better. WDYT?

### Do we want to enforce this behavior regardless of the environment (development/production)?

@irevoire pointed out that it can be frustrating to have this requirement in the development environment for speed and testing reasons. WDYT?

Knowing that instances are exposed to the public with the dev environment, personally, I would be for enforcing it in any case, even if it could be a bump in the DX (when choosing to use a master key in development env) for the sake of security.

Another issue I see with that is that some users may be frustrated to see that it's asked when going for production, aka exposing it to the public (if they change the env to production) -> They will be forced to change the master key and thus have to update every client because the API Keys are updated. IMO that's a super bump.

### CLI error message to indicate that the master key is not secure enough

Note: This is a suggestion

```Error: The provided master key `:givenMasterKey` is not strong enough. It must be composed of `:numberOfCharacters` minimum utf-8 characters.```

### Do we want to give an extra helper to the error message?

If yes, the chosen method should be indicated in the above error message.

- It can be a command the OS provides that generates a random string of utf-8 `:numberOfCharacters`.
- It can be a Meilisearch command that does this.

**Reply by dureuill** [reply]
Thanks for the summary @gmourier!

> We don't necessarily want to wait for a specific format. Asking for a minimum number of characters could be sufficient to guide users to a more secure master key without requiring a format that can be difficult to generate. (@dureuill, please correct me if it's unclear or does not relate to your suggestion).

Regarding this, the rationale is to limit the breakage of this change: we can make three buckets of users:

1. Users with too short master keys
2. Users with long master keys of any format
3. Users with long master keys of the precise format we would decide on

Imposing a format would break both buckets 1 and 2, whereas imposing only a minimum size would break only bucket 1 (which is probably insecure already). This breaking change imposes to change all API keys upon migration, so we would probably prefer not to break people without a clear advantage 

> 16 was mentioned by @dureuill

My understanding was that 16 is needed from the issue that @brunoocasali linked above (https://github.com/meilisearch/meilisearch-dotnet/issues/363#issuecomment-1332593433). In case my understanding is wrong, please correct me, because we're going to break more users with a bigger size requirement (OTOH, 8 seems a bit low with today's available computing power)

>  it can be frustrating to have this requirement in the development environment for speed and testing reasons. WDYT?

Could we describe a bit more the use case of having master keys in dev environments? Looks to me like either we can use it "in development mode" without a master key at all, or we must use it with a secure master key (they can also be generated by and stored in a password manager). Looks error-prone to me to have the constraint lifted only in development contexts, so I think we need a clear advantage to do so.

**Reply by curquiza** [reply]
> How many minimum characters should Meilisearch expect for the master key to be valid?

I said 8 but I have no idea what is the best practice, to be honest. Don't rely on my call for the number of character 😅 

> Do we want to enforce this behavior regardless of the environment (development/production)?

I would always prefer having the same behavior for dev and for production. Can be frustrating to "migrate" from dev to production. Plus, you can easily do tests in dev mode since the master key is not required.

**Reply by dureuill** [reply]
> Do we want to give an extra helper to the error message?

I think this is highly desirable to do. I wonder if, in addition to the `--generate-masterkey` flag, we couldn't go ahead and display one in the error message itself?

```
Error: The provided master key `:givenMasterKey` does not meet the required size of `:numberOfCharacters` characters. Here's a sample master key that is valid for use with Meilisearch: `:generatedMasterKey` (use the `--generate-master-key` command line option to generate other valid keys)
```

?

**Reply by bidoubiwa** [reply]
> We don't necessarily want to wait for a specific format. Asking for a minimum number of characters could be sufficient to guide users to a more secure master key without requiring a format that can be difficult to generate. (@dureuill, please correct me if it's unclear or does not relate to your suggestion).

I 100% agree.

> How many minimum characters should Meilisearch expect for the master key to be valid?

I think 8 might be a bit weak. For me 10 is enough though, but like @curquiza I don't know the best practices. And if like you said @dureuill, with today's machine power 10 is easily cracked, then lets go with 16.

> @irevoire pointed out that it can be frustrating to have this requirement in the development environment for speed and testing reasons. WDYT?

For me, dev mode should be as simple as possible. Most environments provide an easy way to try out. For example, creating a debian server, by default I have an admin account that I can connect to. In SQL, by default most of the tools will provide me with a password `root`. The more we add complexity in trying out Meilisearch, the more we are losing low/middle-level devs or even product teams from other companies that might try out Meilisearch and prepare a quick POC. While @gmourier is a very good dev, I don't think most product owners in non-tech companies (e-commerce etc) are that good.


> CLI error message to indicate that the master key is not secure enough

I love explicit errors 👌


> Do we want to give an extra helper to the error message?

I love simple solutions 👌

**Reply by gmourier** [reply]
Thank you! 

WDYT of that concerns @bidoubiwa? 

"Another issue I see with (**context: not ensuring the master key in the dev environment**) is that some users may be frustrated to see that it's asked when going for production, aka exposing it to the public (if they change the env to production) -> They will be forced to change the master key and thus have to update every client because the API Key keys are updated. IMO that's a super bump."

It seems easier to overcome a message asking to generate a master key of X characters (even more if the error indicates how to do it for newcomers) than to have more important and tedious problems afterward when making the jump to go into production.

"That's a bump.. I just want to try it out. Ok, at least there is a message helping me to do it". Note: The master key is not mandatory in the dev environment.

VS

"Oh... All the API keys have changed. We have to update all client's code right now because the search requests are failing".

**Reply by irevoire** [reply]
Personally, I think that's a wrong concern to have.

> They will be forced to change the master key and thus have to update every client because the API Key keys are updated

The dev mode is supposed to be used for trying out meilisearch and seeing which features you can implement or not. Regenerating your keys before when you go into production seems pretty obvious.

> It seems easier to overcome a message asking to generate a master key of X characters (even more if the error indicates how to do it for newcomers) than to have more important and tedious problems afterward when making the jump to go into production.

My main issue with that is that you may give up on meilisearch when you receive this message first. Not when you are thinking about running meilisearch in production.
So even if it's a bump, people aren't going to leave at this point.

**Reply by gmourier** [reply]
> Regenerating your keys before when you go into production seems pretty obvious.

We have seen users serving meilisearch in production without a master key or a weak master key. This may be the case for experienced dev, but we can't ensure that statement.

> The dev mode is supposed to be used for trying out meilisearch and seeing which features you can implement or not. Regenerating your keys before when you go into production seems pretty obvious.

Yes, but the dev mode does not enforce the usage of a master key. It's a matter of seconds to solve it versus a matter of stressful minutes and maybe hours to update all the client's code because a weak dev master key has to be changed, leading to client code being unable to communicate with Meilisearch at that very particular moment.

I have a hard time understanding how it can be more negative for the overall DX than asking to set a minimum number of characters in case a master key is set by the user in dev env (**knowing that it's not mandatory by default**).

**Reply by bidoubiwa** [reply]
> My main issue with that is that you may give up on meilisearch when you receive this message first. Not when you are thinking about running meilisearch in production.
So even if it's a bump, people aren't going to leave at this point.

I was going to say the same. If I'm looking to go into production with Meilisearch, I might be a bit slowed back I'm not going to drop Meilisearch altogether. Considering also that we are talking about a sub-part of users that did not read the documentation correctly and not all of our users.
While, the more steps I have to launch my first Meilisearch instance, the more chances I'll stop at one of them.

> I have a hard time understanding how it can be more negative for the overall DX than asking to set a minimum number of characters in case a master key is set by the user in dev env (knowing that it's not mandatory by default).

For me, it's common in dev mode to play around without security or external considerations. Just concentrating on the tool itself. In a frictionless onboarding experience. Having a weak master key seems a bit of a detail, but just having the possibility to launch a Meili, and be able to try any request without having to find where I wrote my 16 long char master key and instead just use the one I have in my memory brings way less friction.

**Reply by dureuill** [reply]
> Just concentrating on the tool itself. In a frictionless onboarding experience. Having a weak master key seems a bit of a detail, but just having the possibility to launch a Meili, and be able to try any request without having to find where I wrote my 16 long char master key and instead just use the one I have in my memory brings way less friction.

I re-read the discussion but I guess that, to me, the unclear thing is why a dev trying out the tool in a controlled environment would want to set a master key at all. I think I'm missing some key perspective here that y'all might have gained from experience using the tool. Could you give me some insight into this? 

The main drawback I see about not having the size restriction lifted in dev mode is that we have three different possible behaviors, which sounds complicated and a bit confusing. To list them:
1. dev mode, no master key at all
2. dev mode, master key without a minimum size
3. prod mode, master key with a minimum size

This brings forth another option (option A) that I don't think we discussed yet: have the size issue be a warning rather than an error, in all modes. I see several pros to this solution:

1. No need for three modes of operation, devs that don't bother will simply ignore the warning while in dev
2. No breakage at all, users that have a weak master key will simply see a warning and have the choice to change it when they see fit.

The con however is that we don't always read warnings 🤓

A variant (option B) would be the following behavior:

1. dev mode accepts no master key or a master key without a minimum size, but warns if there is no master key **or** if its size is too small. We can even make the warning specify that the user will need to change the master key and thus all API keys when switching to prod mode.
2. prod mode only accepts a master key with the correct minimum size size

This variant has the pros of making sure no prod environment uses weak master keys while telling users in advance that their current master key is not satisfying the requirements as well as the consequence for prod mode.

What do you think of options A and B?

**Reply by irevoire** [reply]
> the unclear thing is why a dev trying out the tool in a controlled environment would want to set a master key at all.

Yeah, I feel like there is a misconception about that.
When someone tries out meilisearch, he'll obviously first try out the search. But then, before chatting with his team about this new super tool, there is a good chance he's going to check what he can do with the API keys, and if meilisearch can really work for him / his company / integrate with their workflow.

I think @qdequele did a presentation about that some time ago, and he explained something about the « first » onboarding on a product;
1. With meilisearch, the first thing you're going to check is if the search works and if it's the end result you would like to have. That should happen in the first ~10 minutes of usage
2. The second part of the onboarding process that can last for a few hours is when you try to see if meilisearch could fit into your workflow. That's the stage where you're probably going to check if meilisearch can handle a bigger payload, how the tasks queue works if you encounter an issue, and how the API key works.
3. If that's a success, then there is good luck you're going to move on with meilisearch and maybe our cloud.

And I'm pretty sure that when you're at the second part of the onboarding process, you're not using a tool like postman, meili or an sdk to send your request to meilisearch; thus, you need to:
- copy and paste content from the documentation
- copy and paste your API key because it's too complex to memorize

If we allow anything as a master key (for example, `MASTER_KEY` is used everywhere in the documentation and can be copy pasted as-is), we just ease the introduction in meilisearch.

---------

Personally, I'm 100% for the _second_ (B) solution; it eases the access to meilisearch + the move to the production environment.

**Reply by dureuill** [reply]
Thank you for the detailed breakdown of the onboarding process, this is some very useful insight.

Just to clarify what I mean regarding A/B, here's a small table to break it down:


1. **Missing master key**:

| Solution\Environment | Development | Production 
| ----------------------|-------------- |------------
| Current status | Nothing | Error 
| A | Warn | Error
| B | Warn | Error


2. **Too short master key** (new):

| Solution\Environment | Development | Production 
| ----------------------|-------------- |------------
| Current status | Nothing | Nothing
| A | Warn | Warn
| B | Warn | Error

I hope this is clearer :-)

**Reply by ghost** [reply]
Hello all!
Sliding into this conversation to offer a little proposal if you think it might benefit the current predicament.

If you're still quite split over option A or B, there is also the option to try and ask our users on Discord what they think and/or if they have any overall preference on the question?

No matter the decision, I believe it might be very insightful to dedicate a section of the documentation to the question and explain our choice down the line for educational purposes. Just recently, we've met a user that seemed to struggle a little bit with the finer details around master keys and security, so if it can make onboarding a bit smoother it might be worth it. : )

**Reply by dureuill** [reply]
Hey Luna :wave:

if we're asking users, maybe we could share more than just options A and B, but also some of the proposals up thread? Also, some of the broader context of why we're making master keys stronger?

Thinking back on A/B, I have a strong preference for B, because it will reduce the risk of sending weak master keys to prod, and erroring in that case is consistent with what we do in prod when there is no master key at all

**Reply by gmourier** [reply]
Thanks for the solution proposal @dureuill! I am still concerned that the warning will not necessarily be seen 100% of the time and that it does not completely secure the flow expressed above (API keys changing -> clients in failure when switching to the production environment), but I can also see the pros given by @bidoubiwa and @irevoire!

I'm all for @Luna-meili's discord discussion suggestion;  it can help us to see a trend and discuss it with the participants. We must ensure that the proposals enumerate the pros and cons in detail so that community members can project themselves in the best possible way.

**Reply by dureuill** [reply]
OK, so to summarize the first proposal up thread (let's call it C to not change the names of A and B), the updated tables would look this way:


1. **Missing master key**:

| Solution\Environment | Development | Production 
| ----------------------|-------------- |------------
| Current status | Nothing | Error 
| A | Warn | Error
| B | Warn | Error
| C (unchanged from current status) | Nothing | Error 


2. **Too short master key** (new):

| Solution\Environment | Development | Production 
| ----------------------|-------------- |------------
| Current status | Nothing | Nothing
| A | Warn | Warn
| B | Warn | Error
| C | Error | Error

Is that correct @gmourier? If not I'm OK with you providing your own tables

**Reply by gmourier** [reply]
This is correct @dureuill! Thank you 🧱

**Reply by dureuill** [reply]
Some more thoughts that come to mind about A, B, C:

Turning a warning into an error is a breaking change, but the reverse is not true. So we can go from solution C to B to A without breaking back-compatibility, but not from A to B or B to C. If we are hesitating between all three solutions this might be a factor to decide.

**Reply by juchom** [reply]
Another option would be that an instance has always a master key and the behaviours between Development and Production are the same.

It can be auto-generated and displayed on startup (be careful for docker as logs won't be displayed) or supplied by the cli or env variable.

I would go for 16 characters long at least.

I don't really like this idea, but if you want to less frustrate your users you could define a default master key for development environment and give it in the documentations and the samples. But you take the risk of a user misconfiguring Meilisearch and have a development instance exposed with the default key on internet.

It's better to have the same experience on dev and prod and have a good getting started doc to make sure your users will know what to do when going live.

**Reply by dureuill** [reply]
Hey @juchom :wave:

If I summarize these new options, these could be options D and E such that:


1. **Missing master key**:

| Solution\Environment | Development | Production 
| ----------------------|-------------- |------------
| Current status | Nothing | Error 
| A | Warn | Error
| B | Warn | Error
| C (unchanged from current status) | Nothing | Error
| D | Autogenerate | Autogenerate
| E | Use default master-key | Use default master-key

What's not really clear to me is ho Meilisearch should react in options D and E if the user overrides the default/autogenerated master key with a key that doesn't meet the size requirement. Should the default/autogenerated master key override the supplied key in that case? Should there be an error?

> But you take the risk of a user misconfiguring Meilisearch and have a development instance exposed with the default key on internet.

I don't think we're willing to take that risk 😬 

As for forcing a master key (even autogenerated) even in dev mode, I guess the tradeoff is that we'd lose the real convenience to be able to test meilisearch as a new user by sending requests without using a key at all.

Thank you for detailing these ideas, @juchom !

**Reply by juchom** [reply]
>What's not really clear to me is ho Meilisearch should react in options D and E if the user overrides the default/autogenerated master key with a key that doesn't meet the size requirement.

If the key is not strong enough it is rejected and the original master key is not changed or if it is at the initialization of the server, the server does not start.

> As for forcing a master key (even autogenerated) even in dev mode, I guess the tradeoff is that we'd lose the real convenience to be able to test meilisearch as a new user by sending requests without using a key at all.

Displaying the key like @brunoocasali showed earlier and having a good quick start guide should be fine I think. It's a bit painful for users the first time to deal with a key, but they will have to do it anyway for other cloud services and they will be aware of the importance of a master key.

**Reply by dureuill** [reply]
Thank you for the precisions! My understanding is that the development mode is dedicated to onboarding users in the simplest possible way, and the simplest way is to continue allowing no master key in this mode. While having great documentation is an invaluable asset, especially for those users who need to understand all requirements before jumping in, I think being able to use meilisearch for the first time without reading the documentation related to master keys is invaluable too.

I think I'll provide an implementation of solution B in the next weeks, as I'd really like how visible the warnings in development mode will be in practice. It should be noted that converting the warning into an error in the development mode, that is, switching to solution C, is not really costly implementation-wise, meaning we can probably switch from one solution to the other or even test both solutions in the lead-up to v1. 

Many thanks to everyone involved in this discussion so far 🙌

**Reply by brunoocasali** [reply]
@dureuill, which kind of warning are you thinking about implementing?

Mentioning that, because I consider the current behavior a warning:

<img width="1100" alt="image" src="https://user-images.githubusercontent.com/4116980/208981551-22e2819b-17c2-4a54-bbd8-a0e053c25ddf.png">

**Reply by dureuill** [reply]
Hey @brunoocasali, you can see what I mean in the associated PR: https://github.com/meilisearch/meilisearch/pull/3274

I put a screenshot of each situation. I took the liberty to request a review from you for this PR so that you can tell me what you think of these warnings. Feel free to suggest changes!

**Comment by gmourier** [comment]
Hey everyone 👋

I'm locking that discussion since it will be released for `v1`

Please open a new discussion to share feedback or ask for a feature request.

Thank you all for your hard work 🙇‍♂️
